articles/react-error-handling.md

```diff
@@ -0,0 +1,436 @@
+
+```cmd
+npm install react-error-boundary
+```
+
+After installing the package we need to import the `ErrorBoundary` component and wrap our `renderContent` function in it
+
+```tsx
+import { ErrorBoundary } from 'react-error-boundary'
+
+// rest of the code
+
+<ErrorBoundary fallback={<p>There was an error</p>}>
+  {renderContent()}
+</ErrorBoundary>
+```
+
+![product-search-error-boundary-fallback.gif](/images/articles/react-error-handling/product-search-error-boundary-fallback.gif)
+
+But wait, the uncaught error alert is still being thwron. Is'not this the problem we will solve? Yeah, it is. We are at development mode (using the react's dev libraries) so react throws that error by default it can not be hidden what it did change is the result after closing that alert, our component is still there and we can keep interacting somehow (at production the error alert is not shown, the benefits of ErrorBoundary are way more appreciable).
+
+Before implementing ErrorBoundary if our component crashed the error alert would be shown and the page gets empty. There is a big difference in terms of how our component behaves now.
+
+Before taking ErrorBoundary further let's make a tiny change in our `makeRequest` function so our function will return wrong data once, for the rest of the request the data will be right.
+
+```ts
+const products = {
+  '1': {
+    productName: 'xxxx',
+    price: 0
+  },
+  '2': {
+    productName: 'xxxx',
+    price: 0
+  }
+} as unknown as Record<string, ProductData>
+
+function makeRequestClosure() {
+  let count = 0
+
+  return async function (productId: string): Promise<ProductData> {
+    const product = products[productId]
+  
+    return new Promise((resolve, reject) => {
+      setTimeout(() => {
+        if (product) {
+          
+          if (count > 0) {
+            product.productReference = `${productId}-xxxx`
+          }
+
+          resolve(product)
+          count++
+        } else {
+          reject(new Error('It was not product found'))
+        }
+
+      }, 500)
+    })
+  }
+}
+
+const makeRequest = makeRequestClosure()
+
+const productDetails: Record<string, Record<string, string>> = {
+  '1-xxxx': {
+    care: 'Do not wash in washing machine'
+  },
+  '2-xxxx': {
+    care: 'Do not wash in washing machine'
+  }
+}
+```
+
+We just created the makeRequestClosure with a variable count initialized in 0 and it will increment its value every time a product is found. If count is greater than 0 the productReference is added to the product and the promise is inmediately resolved (If you want to dive deep into closure please at this [link](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures) or let us know if you want us to do an article about it).
```

Comment:
```suggestion
We just created the makeRequestClosure with a variable count initialized at 0 and it will increment its value every time a product is found. If count is greater than 0 the productReference is added to the product and the promise is immediately resolved (If you want to dive deep into closures please at this [link](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures) or let us know if you want us to do an article about it).
```